File: frontend/src/app/core/validators.ts
// Mismo algoritmo que backend/asistencia/validators.py (módulo 10, coeficientes del
// INEC) — feedback inmediato mientras se escribe, el backend sigue siendo la
// validación real al enviar el formulario.
const COEFICIENTES = [2, 1, 2, 1, 2, 1, 2, 1, 2];

export function cedulaEcuatorianaValida(cedula: string): boolean {
  if (!/^\d{10}$/.test(cedula)) return false;

  const provincia = Number(cedula.slice(0, 2));
  const tercerDigito = Number(cedula[2]);
  if (provincia < 1 || provincia > 24 || tercerDigito >= 6) return false;

  let total = 0;
  for (let i = 0; i < 9; i++) {
    const producto = Number(cedula[i]) * COEFICIENTES[i];
    total += producto > 9 ? producto - 9 : producto;
  }

  const verificador = (10 - (total % 10)) % 10;
  return verificador === Number(cedula[9]);
}

/** Deja solo dígitos y corta al largo máximo — para inputs de cédula/teléfono. */
export function soloDigitos(valor: string, largoMaximo: number): string {
  return valor.replace(/\D/g, '').slice(0, largoMaximo);
}

File: backend/asistencia/validators.py
from django.core.exceptions import ValidationError

# Algoritmo oficial de cédula ecuatoriana (módulo 10, coeficientes del INEC).
_COEFICIENTES = (2, 1, 2, 1, 2, 1, 2, 1, 2)


def validar_cedula_ecuatoriana(cedula: str) -> None:
    if not cedula.isdigit() or len(cedula) != 10:
        raise ValidationError("La cédula debe tener 10 dígitos.")

    provincia = int(cedula[0:2])
    tercer_digito = int(cedula[2])
    if not (1 <= provincia <= 24) or tercer_digito >= 6:
        raise ValidationError("La cédula no tiene un formato válido.")

    total = 0
    for digito, coeficiente in zip(cedula[:9], _COEFICIENTES):
        producto = int(digito) * coeficiente
        total += producto - 9 if producto > 9 else producto

    verificador = (10 - total % 10) % 10
    if verificador != int(cedula[9]):
        raise ValidationError("La cédula no es válida (dígito verificador incorrecto).")
